Metavars vector when nargs > 1

samkohn opened the pull request:
When `nargs > 1` for optional arguments, a distinct metavar can be specified for each argument using a `Vector` of `AbstractStrings`.

This is half of what I suggested in #29. The other half is to do the same for argument types.

I am happy to squash my commits but I left them separate here for inspection. I'm also happy to wait to merge until I finish the second half of my self-assigned task. That said, this is a good logical unit to merge in on its own.

carlobaldassi commented:
Thanks, I'll review this sometime this week, as soon as I find the time.

carlobaldassi commented:
Ok so I had a look and it seems mostly good. I have some comments:

1. There doesn't seem to be a check that the length of the vector is the same as the number of expected arguments. This will result e.g. in a crash if the vector is too short, or in unused entries if it is too long. I'd rather give a clear error message.

2. I'm not sure how this could work or even if it should work with positional arguments. The issue here is that in the auto-generated help (and throughout the code, really) the positional arguments are *identified* by the metavar. Also, if your positional arguments have two different meanings, you should probably use two arguments. As far as I can see, this could only work and make sense for *optional* positional arguments, which is perhaps a narrow enough case to ignore. I think overall it is safer not have this feature for positional args for now.

3. It seems like  [this](https://github.com/carlobaldassi/ArgParse.jl/pull/33/files#diff-494b35a6ec809bfb75105d68ac52e3baR1328) is intended to allow the feature for positional arguments, while currently [this error](https://github.com/carlobaldassi/ArgParse.jl/pull/33/files#diff-494b35a6ec809bfb75105d68ac52e3baR787) only allows the feature for options. I'd rather remove the first part for the time being (see point 2).

4. Documentation needs updating.

Squashing the commits seems a good idea.

I'm not sure what's the issue with the failing tests. I wouldn't worry about the Julia 0.3 failures, we'll just remove support. The one about Julia nightly is more troubling since I have not idea of the reason for that, but we'll see, I don't think it's related to this PR.

As an unrelated aside, while reviewing I also noticed a problem in the code related to required options, which I'll need to fix, but I'll keep that for a later time.

samkohn commented:
Updates:

- Checks for `length(metavars) == nargs` (and included a new test case)

- Added a test case for trying to use a metavars Vector with positional arguments (should fail, as you suggest)

- The code you mentioned in your point 3 as allowing positional args is part of the help string generation. The code would never get that far with positional args since an exception would be raised when the arg field is created. I hope the new test case clarifies my intentions.

- I updated the documentation but I am having issues building the Sphinx documents. I just modified the text in one place so there shouldn't be any major issues. But, I guess I'd suggest you check on your end to make sure everything renders correctly.

carlobaldassi commented:
Ok looks good thanks!
